AP_Stats: update flight time on disarm

tatsuy opened the pull request:
This PR fixes an issue where flight time was not being updated if the power was shut down immediately after flight.
Since the update occurs in 30-second cycles, turning off the power before the update process means the last flight time is not added.

I have modified update function so that the flight time updates upon disarming, in addition to the regular updates every 30 seconds.

I have confirmed that the flight time is updated immediately after disarming in SITL.

IamPete1 commented:
Why not add the flush to the `set_flying` call? That is when the flight time will stop counting up so there should be no change in value. The key thing it needs is the flush to save the value to eeprom.

I think you could just add `flush()` in there, also updating `last_flush_ms`.

https://github.com/ArduPilot/ardupilot/blob/61b3ad326dda41449bbd0f954cea126b19b3028a/libraries/AP_Stats/AP_Stats.cpp#L134-L135

It just seems a little odd to trigger the save on arming when the stat is based on flying.

tatsuy commented:
@IamPete1 Thank you for your suggestion.
> I think you could just add flush() in there, also updating last_flush_ms.
In this case, runtime will no longer be updated because update_runtime() is no longer called during disarm

If we add update_runtime() and flush() in there, also updating last_flush_ms, runtime is updated whenever the drone is not flying. It could lead to frequent updates of the runtime value.

I thought it would be better to update once only when disarming.

IamPete1 commented:
> If we add update_runtime() and flush() in there, also updating last_flush_ms, runtime is updated whenever the drone is not flying. It could lead to frequent updates of the runtime value.

Right, you would have to check `_flying_ms`

```
if (_flying_ms) {
    flush();
    last_flush_ms = now;
}
```
Having broken out the millis fetch from the other branch of the if to populate "now".

muramura commented:
This process adds less than 30 seconds to the time.

The notification to GCS is in 30-second increments.
The GCS displays the event as 30 seconds short.


void AP_Stats::set_flying(const bool is_flying)
{
    if (is_flying) {
        if (!_flying_ms) {
            _flying_ms = AP_HAL::millis();
        }
    } else {
        update_flighttime();
        _flying_ms = 0;
    }
}

peterbarker commented:
Concur with Peter and Muramura here; move the update of `last_flash_ms` into the `flush` method itself, however, rather than setting it after calling `flush`

tatsuy commented:
Thank you for the reviews.
I added the `flush` to the `set_flying` call and moved the update of `last_flash_ms` into the `flush` method.
Additionally, I've called not only `update_flighttime` but also `update_runtime` to maintain consistency.

For the test:
Twice, I repeated the process of letting SITL fly and then immediately terminating SITL after disarming.
This PR:
![image](https://github.com/ArduPilot/ardupilot/assets/16643069/7f96eaac-d9c3-4498-bf56-45788cece7b2)
It works.

In case of no `update_runtime()`:
![image](https://github.com/ArduPilot/ardupilot/assets/16643069/dccf0377-eadb-4521-82f3-40bd4d892f24)
STAT_FLTTIME >STAT_RUNTIME and it is not consistent.

peterbarker reviewed (approved):
LGTM
